### DISCUSSION TITLE
Imagining New Usage

### DISCUSSION THREAD

**Discussion by haz** [AI-Planning/pddl#34] [Ideas]
Would like to start a thread on ideas of how the library could be used for planning research. Some of this may be out of scope for the project, but much of it may be viable. We shall see...

**Comment by haz** [comment]
## Custom PDDL

Would be nice to handle custom PDDL syntax for one-off research initiatives. Some examples...

* Nested belief / desire modalities: `(believes ?ag1 (wants ?ag2 (believes ?ag3 (flight-is-on-time))))`
* Labeled non-determinism: `(labeled-oneof my-nondet-effect (outcome happy (and (...))) (outcome sad (and (...))))`

### Usage
Maybe doable by...
1. Making sure the syntax is in addition to what's already offered
2. Inherit from `DomainTransformer` and define some custom lark
3. Pass the new class in as the constructor to `DomainParser` to be used [[here](https://github.com/AI-Planning/pddl/blob/master/pddl/parser/domain.py#L248)]

Will update this usage to the actual code once I've thought it through some more.

**Comment by haz** [comment]
## Domain Transformations

1-liner transformations to STRIPS, non-condeffs (in either of the two common methods), normalized effects (non-det becoming a big flat list of oneof's), etc.

* Should pull in objects to be constants as needbe (for removing existential/universal)
* Should work on the lifted representation as much as possible (grounding can be deadly)
* Should be seen as a standard preprocessor that could be used to compile more expressive formalisms into simpler ones
* Bonus marks if it can handle object fluents (this _is_ the unquestionable PDDL 3.1 parser, n'est-ce pas?)

**Comment by tom4everitt** [comment]
Hi,

It wasn't entirely clear to me from the documentation, whether this package supports the reading of parsing of a domain file and problem file, and get out "task" object with functions telling you
* the current state
* which actions are available in the current state
* allows you to perform one of the actions, and get an updated state
* check if the current state is a goal state

This would the package useful for planning / search applications.

**Reply by haz** [reply]
Ya, I think those elements are out of scope for what this package is aimed at. It's more about the parse, create, modify, write. Action mechanics (applicability, progression, regression, entailment, etc.) would be for a library that _uses_ `pddl`. You might want to look at `tarski` for something that does grounding and progression. I plan on writing something with the functionality you're looking for, built on top of `pddl`, but that might be a ways in the future.

**Reply by haz** [reply]
Just as an update, I've finally gotten around to starting such a wrapper library: https://github.com/AI-Planning/pddl-utils

**Comment by tom4everitt** [comment]
thanks, I've already found another workaround, but maybe someone else will
find it useful

On Sat, Jan 4, 2025 at 5:27 PM Christian Muise ***@***.***>
wrote:

> Just as an update, I've finally gotten around to starting such a wrapper
> library: https://github.com/AI-Planning/pddl-utils
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AI-Planning/pddl/discussions/34#discussioncomment-11734613>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEIABRAP7S5GMIDU7ZAGML2JAKZPAVCNFSM6AAAAABUTI5FOKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCNZTGQ3DCMY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Tom
tomeveritt.se
